From a2bc4d72fe6c8e888534fb4049374af4fef394e5 Mon Sep 17 00:00:00 2001
Message-Id: <a2bc4d72fe6c8e888534fb4049374af4fef394e5.1497444332.git.maciej.borzecki@rndity.com>
In-Reply-To: <cover.1497444332.git.maciej.borzecki@rndity.com>
References: <cover.1497444332.git.maciej.borzecki@rndity.com>
From: Maciej Borzecki <maciej.borzecki@rndity.com>
Date: Mon, 12 Jun 2017 23:14:40 +0200
Subject: [PATCH 1/5] vexpress_ca9x4: enable booting from ubifs

Enable options that add support for UBI images on CFI and UBIFS.

Enable MTD concatenation, thus 2*64MB NOR flash appears as single 128MB device.

Define the follwing flash partitions:
 - 1MB - uboot
 - 1MB - uboot environment (previously stored at the top of the flash)
 - remaining - UBI

When defining mtdparts use device name that matches kernel's MTD name (otherwise
partitioning will not be applied).

Signed-off-by: Maciej Borzecki <maciej.borzecki@rndity.com>
---
 configs/vexpress_ca9x4_defconfig |  5 ++++
 include/configs/vexpress_ca9x4.h | 57 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)

diff --git a/configs/vexpress_ca9x4_defconfig b/configs/vexpress_ca9x4_defconfig
index 7b845c6c79843f99b04bc64d80d2e13d6a53b494..554e07b4c25ad4ec7b469c9e8f830da4404eca6a 100644
--- a/configs/vexpress_ca9x4_defconfig
+++ b/configs/vexpress_ca9x4_defconfig
@@ -20,3 +20,6 @@ CONFIG_CMD_MMC=y
 CONFIG_MTD_NOR_FLASH=y
 CONFIG_BAUDRATE=38400
 CONFIG_OF_LIBFDT=y
+CONFIG_CMD_UBI=y
+CONFIG_CMD_UBIFS=y
+CONFIG_CMD_MTDPARTS=y
diff --git a/include/configs/vexpress_ca9x4.h b/include/configs/vexpress_ca9x4.h
index 993398ccc696c70d7f4c23c13f00fda6c0f536fe..4138a0e0ff18ff138bfe4918cd75a3494366d94b 100644
--- a/include/configs/vexpress_ca9x4.h
+++ b/include/configs/vexpress_ca9x4.h
@@ -14,4 +14,61 @@
 #define CONFIG_VEXPRESS_ORIGINAL_MEMORY_MAP
 #include "vexpress_common.h"
 
+#define CONFIG_RBTREE
+#define CONFIG_CMD_MTDPARTS
+#define CONFIG_MTD_DEVICE
+#define CONFIG_MTD_PARTITIONS
+#define CONFIG_MTD_CONCAT
+#define CONFIG_FLASH_CFI_MTD
+#define CONFIG_LZO
+
+#ifdef CONFIG_SYS_MALLOC_LEN
+#undef CONFIG_SYS_MALLOC_LEN
+#define CONFIG_SYS_MALLOC_LEN (8 * 1024 * 1024)
+#endif
+
+/* CONFIG_MTD_CONCAT is set, hence nor0 and nor1 will become a single device
+   named nor2
+
+   NOTE: NOR flash appears as a single MTD device with name 40000000.flash */
+#define MTDIDS_DEFAULT		"nor2=40000000.flash"
+#define MTDPARTS_DEFAULT  "mtdparts=40000000.flash:"  \
+  "1m(u-boot)ro," \
+  "1m(u-boot-env)ro," \
+  "-(ubi)"
+
+#ifdef CONFIG_EXTRA_ENV_SETTINGS
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#endif
+
+#define UBI_BOOTCMD \
+ 	"ubiboot=" \
+	"echo Booting from NOR...; "                                          \
+	"ubi part ubi && "                                                    \
+	"ubifsmount ubi0:rootfs && "                                          \
+	"ubifsload ${kernel_addr_r} /boot/zImage && "                         \
+	"ubifsload ${fdt_addr_r} /boot/${fdtfile} && "                        \
+  "setenv bootargs ${mtdparts} ${ubiargs} ${defargs} && "               \
+	"bootz ${kernel_addr_r} - ${fdt_addr_r}\0"
+
+#define CONFIG_EXTRA_ENV_SETTINGS                                       \
+  UBI_BOOTCMD \
+  "kernel_addr_r=0x60100000\0"                                          \
+  "fdt_addr_r=0x60000000\0"                                             \
+  "fdtfile=vexpress-v2p-ca9.dtb\0"                                      \
+  "mtdparts=" MTDPARTS_DEFAULT "\0"                                     \
+  "ubiargs=ubi.mtd=2 root=ubi0:rootfs rootfstype=ubifs ubi.fm_autoconvert=1\0" \
+  "defargs=console=ttyAMA0,115200n8 panic=3\0"
+
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND "run ubiboot; reset; "
+
+/* environment starts at 1MB offset in the flash */
+#undef CONFIG_ENV_OFFSET
+#define CONFIG_ENV_OFFSET 0x100000
+/* was fixed in vexpress_common, but will be derived in environment.h */
+#undef CONFIG_ENV_ADDR
+/* required for automatic calculation of CONFIG_ENV_ADDR */
+#define CONFIG_SYS_FLASH_BASE CONFIG_SYS_FLASH_BASE0
+
 #endif /* VEXPRESS_CA9X4_H */
-- 
2.9.3

